src/i_joystick.c

```diff
@@ -76,6 +81,246 @@ static int joystick_physical_buttons[NUM_VIRTUAL_BUTTONS] = {
     0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10
 };
 
+void I_ShutdownGamepad(void)
+{
+    if (gamepad != NULL)
+    {
+        SDL_GameControllerClose(gamepad);
+        gamepad = NULL;
+        SDL_QuitSubSystem(SDL_INIT_GAMECONTROLLER);
+    }
+}
+
+static int FindFirstGamepad(void)
+{
+    int i;
+    int gamepadindex = -1;
+
+    for (i = 0; i < SDL_NumJoysticks(); ++i)
+    {
+        if (SDL_IsGameController(i))
+        {
+            gamepadindex = i;
+            break;
+        }
+    }
+
+    return gamepadindex;
+}
+
+static int FindSpecificGamepad(SDL_JoystickGUID guid)
+{
+    SDL_JoystickGUID dev_guid;
+    int i;
+    int gamepadindex = -1;
+
+    for (i = 0; i < SDL_NumJoysticks(); ++i)
+    {
+        dev_guid = SDL_JoystickGetDeviceGUID(i);
+        if (!memcmp(&guid, &dev_guid, sizeof(SDL_JoystickGUID)))
+        {
+            gamepadindex = i;
+            break;
+        }
+    }
+
+    return gamepadindex;
+}
+
+static int DeviceIndexGamepad(void)
+{
+    SDL_JoystickGUID guid, dev_guid;
+    int index = -1;
+
+    if (strcmp(joystick_guid, ""))
+    {
+        guid = SDL_JoystickGetGUIDFromString(joystick_guid);
+
+        // First, look for the gamepad at the previously-used index.
+        if (joystick_index >= 0 && joystick_index < SDL_NumJoysticks())
+        {
+            dev_guid = SDL_JoystickGetDeviceGUID(joystick_index);
+            if (!memcmp(&guid, &dev_guid, sizeof(SDL_JoystickGUID)))
+            {
+                return joystick_index;
+            }
+        }
+
+        // Maybe the index has moved?
+        index = FindSpecificGamepad(guid);
+    }
+
+    // Find a gamepad for the first time. Or, if the previous gamepad isn't
+    // present, see if a different one is available.
+    if (index < 0)
+    {
+        index = FindFirstGamepad();
+    }
+
+    return index;
+}
+
+void I_InitGamepad(void)
+{
+    SDL_JoystickGUID guid;
+    int index;
+
+    if (!use_gamepad)
+    {
+        return;
+    }
+
+    if (SDL_Init(SDL_INIT_GAMECONTROLLER) < 0)
+    {
+        return;
+    }
+
+    index = DeviceIndexGamepad();
+
+    if (index < 0)
+    {
+        printf("I_InitGamepad: No gamepad found.\n");
+        SDL_QuitSubSystem(SDL_INIT_GAMECONTROLLER);
+        return;
+    }
+
+    gamepad = SDL_GameControllerOpen(index);
+
+    if (gamepad == NULL)
+    {
+        printf("I_InitGamepad: Failed to open gamepad.\n");
+        SDL_QuitSubSystem(SDL_INIT_GAMECONTROLLER);
+        return;
+    }
+
+    joystick_index = index;
+
+    if (strcmp(joystick_guid, ""))
+    {
+        joystick_guid = malloc(33);
```

Comment:
Magic number, use sizeof instead

Reply:
The 33 comes from `SDL_JoystickGetGUIDString` documentation (I don't think there's anything you can sizeof to get this) but I definitely agree it should be deobfuscated.

Reply:
33eac07